lib/Doctrine/ODM/MongoDB/Proxy/Factory/LazyGhostProxyFactory.php

```diff
@@ -0,0 +1,385 @@
+        $filter            = ReflectionProperty::IS_PUBLIC | ReflectionProperty::IS_PROTECTED | ReflectionProperty::IS_PRIVATE;
+        $reflector         = $class->getReflectionClass();
+
+        while ($reflector) {
+            foreach ($reflector->getProperties($filter) as $property) {
+                $name = $property->name;
+
+                if ($property->isStatic() || (($class->hasField($name) || $class->hasAssociation($name)))) {
+                    continue;
+                }
+
+                $prefix = $property->isPrivate() ? "\0" . $property->class . "\0" : ($property->isProtected() ? "\0*\0" : '');
+
+                $skippedProperties[$prefix . $name] = true;
+            }
+
+            $filter    = ReflectionProperty::IS_PRIVATE;
+            $reflector = $reflector->getParentClass();
+        }
+
+        $className       = $class->getName(); // aliases and case sensitivity
+        $entityPersister = $this->uow->getDocumentPersister($className);
+        $initializer     = $this->createLazyInitializer($class, $entityPersister);
+        $proxyClassName  = $this->loadProxyClass($class);
+
+        $proxyFactory = Closure::bind(static function (mixed $identifier) use ($initializer, $skippedProperties, $class): InternalProxy {
+            /** @see LazyGhostTrait::createLazyGhost() */
+            $proxy = static::createLazyGhost(static function (InternalProxy $object) use ($initializer, $identifier): void {
+                $initializer($object, $identifier);
+            }, $skippedProperties);
+
+            $class->setIdentifierValue($proxy, $identifier);
+
+            return $proxy;
+        }, null, $proxyClassName);
+
+        return $this->proxyFactories[$className] = $proxyFactory;
+    }
+
+    private function loadProxyClass(ClassMetadata $class): string
+    {
+        $proxyClassName = self::generateProxyClassName($class->getName(), $this->proxyNs);
+
+        if (class_exists($proxyClassName, false)) {
+            return $proxyClassName;
+        }
+
+        if ($this->autoGenerate === Configuration::AUTOGENERATE_EVAL) {
+            $this->generateProxyClass($class, null, $proxyClassName);
+
+            return $proxyClassName;
+        }
+
+        $fileName = $this->getProxyFileName($class->getName(), $this->proxyDir);
+
+        switch ($this->autoGenerate) {
+            case Configuration::AUTOGENERATE_FILE_NOT_EXISTS_OR_CHANGED:
+                if (file_exists($fileName) && filemtime($fileName) >= filemtime($class->getReflectionClass()->getFileName())) {
+                    break;
+                }
+            // no break
+            case Configuration::AUTOGENERATE_FILE_NOT_EXISTS:
+                if (file_exists($fileName)) {
+                    break;
+                }
+            // no break
+            case Configuration::AUTOGENERATE_ALWAYS:
+                $this->generateProxyClass($class, $fileName, $proxyClassName);
+                break;
+        }
```

Comment:
It would be more intuitive this way:
```suggestion
        if (
            match ($this->autoGenerate) {
                Configuration::AUTOGENERATE_FILE_NOT_EXISTS_OR_CHANGED => ! file_exists($fileName) || filemtime($fileName) < filemtime($class->getReflectionClass()->getFileName()),
                Configuration::AUTOGENERATE_FILE_NOT_EXISTS => ! file_exists($fileName),
                Configuration::AUTOGENERATE_ALWAYS => true,
                Configuration::AUTOGENERATE_NEVER => false,
            }
        ) {
            $this->generateProxyClass($class, $fileName, $proxyClassName);
        }
```